File: matlab/lib/kpctoolbox/map/rap_sample.m
function sample = rap_sample(RAP, n)
% SAMPLE = RAP_SAMPLE(RAP, N) - Generate a correlated sample path from a RAP
%
% Generate N successive inter-event times along one sample path of a
% Rational Arrival Process. A RAP has no underlying Markov chain over the
% phases to walk, so the sampler carries the conditional phase vector V
% across events instead of a phase index.
%
% Input:
%   RAP: RAP distribution as either:
%        - RAP object
%        - Process cell array {H0, H1}
%   N: Number of samples to generate (default: 1)
%
% Output:
%   SAMPLE: Column vector of N inter-event times along a single sample path
%
% Algorithm:
%   Let V be a row vector with V*e = 1, initialized to the arrival-embedded
%   equilibrium vector map_pie. For each draw:
%     1. Draw U ~ Uniform(0,1) and solve the conditional survival equation
%        S(X) = V*expm(H0*X)*e = 1-U for X, by bracket expansion followed by
%        safeguarded Newton iteration. S is monotone decreasing from 1 to 0,
%        and its derivative is S'(X) = V*expm(H0*X)*H0*e, so the conditional
%        density is -S'(X) = V*expm(H0*X)*H1*e.
%     2. Update V <- V*expm(H0*X)*H1 / (V*expm(H0*X)*H1*e).
%
%   Step 2 is what reproduces the autocorrelation of the process. Sampling
%   from the ME marginal alone (me_sample) yields independent inter-event
%   times with the correct marginal but zero autocorrelation, and the MAP
%   CTMC walk (map_sample) is invalid here because a general RAP has
%   negative off-diagonal entries, for which the walk has no probabilistic
%   interpretation. A MAP is the special case in which H0 and H1 are
%   nonnegative, and on that input this sampler agrees with map_sample.
%
% Examples:
%   rap = RAP([-2, 1; 0.5, -1.5], [0.5, 0.5; 0.5, 0.5]);
%   samples = rap_sample(rap, 10000);
%
%   % Or use process representation directly
%   RAP_proc = {[-2, 1; 0.5, -1.5], [0.5, 0.5; 0.5, 0.5]};
%   samples = rap_sample(RAP_proc, 10000);
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

if nargin < 2
    n = 1;
end

% Extract process representation
if isa(RAP, 'RAP')
    RAP_proc = RAP.getProcess();
elseif iscell(RAP)
    RAP_proc = RAP;
else
    error('RAP must be either a RAP object or a cell array {H0, H1}');
end

H0 = RAP_proc{1};
H1 = RAP_proc{2};
nphases = size(H0, 1);
e = ones(nphases, 1);

% Conditional phase vector, normalized so that v*e = 1. The arrival-embedded
% equilibrium vector is the stationary choice for the first inter-event time.
v = map_pie(RAP_proc);
v = v / (v * e);

% Scale for the initial bracket and for the Newton safeguard.
meanval = map_mean(RAP_proc);
if ~isfinite(meanval) || meanval <= 0
    meanval = 1;
end

sample = zeros(n, 1);
for i = 1:n
    target = 1 - rand();  % target survival level in (0,1]

    x = invert_survival(v, H0, e, target, meanval);
    sample(i) = x;

    % Advance the conditional vector across the event.
    vnext = v * expm(H0 * x) * H1;
    mass = vnext * e;
    if mass <= 0 || ~isfinite(mass)
        % The conditional vector has lost its normalization to roundoff,
        % which can only happen at survival levels far into the tail. Restart
        % from the embedded equilibrium rather than propagate a meaningless
        % vector.
        v = map_pie(RAP_proc);
        v = v / (v * e);
    else
        v = vnext / mass;
    end
end

end

function x = invert_survival(v, H0, e, target, scale)
% Solve v*expm(H0*x)*e = target for x, with target in (0,1].

surv = @(t) v * expm(H0 * t) * e;

% Bracket the root: survival is 1 at t = 0 and decreases to 0.
lo = 0;
hi = scale;
smax = 200;
k = 0;
while surv(hi) > target && k < smax
    lo = hi;
    hi = hi * 2;
    k = k + 1;
end
if k >= smax
    x = hi;
    return;
end

% Bisection to a tight bracket, then Newton polish. Bisection alone is used
% first because the conditional density of a RAP need not be monotone, so an
% unguarded Newton step from an arbitrary start can leave the bracket.
for k = 1:60
    mid = 0.5 * (lo + hi);
    if surv(mid) > target
        lo = mid;
    else
        hi = mid;
    end
    if (hi - lo) <= 1e-12 * max(1, hi)
        break;
    end
end

x = 0.5 * (lo + hi);
for k = 1:3
    ex = expm(H0 * x);
    s = v * ex * e;
    d = v * ex * H0 * e;   % derivative of the survival function, negative
    if ~(d < 0)
        break;
    end
    xn = x - (s - target) / d;
    if xn <= lo || xn >= hi
        break;
    end
    if abs(xn - x) <= 1e-14 * max(1, abs(x))
        x = xn;
        break;
    end
    x = xn;
end

end

File: jar/src/main/java/jline/api/mam/Rap_sample.java
package jline.api.mam;

import java.util.Random;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

/**
 * Rational Arrival Process (RAP) sampling by conditional inversion.
 *
 * <p>A RAP (H0, H1) generalises a MAP by dropping the sign constraints on the
 * two matrices, so the CTMC walk of {@link Map_sample} does not apply. Sampling
 * is instead performed on the conditional (vector) representation: with a row
 * vector v satisfying {@code v*e = 1}, the next inter-event time has survival
 * function {@code S_v(x) = v*expm(H0*x)*e} and density
 * {@code f_v(x) = -v*expm(H0*x)*H0*e = v*expm(H0*x)*H1*e}, and after an event at
 * x the vector is updated to</p>
 *
 * <pre>{@code v <- v*expm(H0*x)*H1 / (v*expm(H0*x)*H1*e)}</pre>
 *
 * <p>The generated sequence is therefore <em>correlated</em>: it reproduces the
 * autocorrelation of the process, not merely its ME marginal. Starting v from
 * the embedded stationary vector {@link Map_pie} makes each inter-event time
 * marginally stationary. A MAP is the special case in which H0 has non-negative
 * off-diagonal entries and H1 is non-negative; the algorithm is exact there too,
 * although {@link Map_sample.MapSampler} is cheaper for that case.</p>
 *
 * <p>Because v changes at every draw, no inversion table can be cached across
 * draws. What is cached is the sequence {@code expm(H0*t_i)} on a fixed time
 * grid (built with a single matrix exponential and repeated matrix products),
 * together with the column vectors {@code expm(H0*t_i)*e}. A draw then costs
 * only vector products: the conditional survival at a grid point is a dot
 * product against the cached column, so no matrix exponential is evaluated per
 * draw.</p>
 *
 * @since LINE 3.0
 */
public final class Rap_sample {
    private Rap_sample() {}

    /**
     * Stateful RAP sampler that carries the conditional phase vector across
     * draws, so successive inter-event times are correlated as in the process.
     */
    public static final class RapSampler {
        private final int n;
        private final Matrix d0Matrix;
        private final double[][] d0;
        private final double[][] d1;
        private final double[] rowSumD0;
        private final double normD0;
        private final double[] pie;
        private double[] v;
        private final int gridSize;
        private final double h;
        private final double[] tGrid;
        private final double[][][] eGrid;
        private final double[][] gGrid;
        private final double tEnd;
        private final double eta;
        private final boolean exponential;
        private final double expRate;

        /**
         * Builds the cached grid of matrix exponentials of H0 and initialises the
         * conditional vector to the embedded stationary vector.
         *
         * @param D0 the hidden-transition matrix H0
         * @param D1 the visible-transition matrix H1
         */
        public RapSampler(Matrix D0, Matrix D1) {
            this.n = D0.getNumRows();
            this.d0Matrix = D0;
            this.d0 = Me_sample.toArray2D(D0);
            this.d1 = Me_sample.toArray2D(D1);
            this.rowSumD0 = new double[n];
            double maxRowAbs = 0.0;
            for (int i = 0; i < n; i++) {
                double s = 0.0;
                double abs = 0.0;
                for (int j = 0; j < n; j++) {
                    s += d0[i][j];
                    abs += Math.abs(d0[i][j]);
                }
                rowSumD0[i] = s;
                if (abs > maxRowAbs) maxRowAbs = abs;
            }
            this.normD0 = maxRowAbs;

            boolean isExp = (n == 1) && d0[0][0] < 0.0;
            this.exponential = isExp;
            this.expRate = isExp ? -d0[0][0] : 0.0;

            this.pie = new double[n];
            if (isExp) {
                this.pie[0] = 1.0;
                this.v = this.pie.clone();
                this.gridSize = 0;
                this.h = 0.0;
                this.tGrid = null;
                this.eGrid = null;
                this.gGrid = null;
                this.tEnd = 0.0;
                this.eta = -expRate;
                return;
            }

            Matrix pieMatrix = Map_pie.map_pie(D0, D1);
            for (int j = 0; j < n; j++) {
                this.pie[j] = pieMatrix.get(0, j);
            }
            this.v = this.pie.clone();

            double mean = Map_mean.map_mean(D0, D1);
            double var = Map_var.map_var(D0, D1);
            double sigma = (var > 0.0 && !Double.isNaN(var)) ? Math.sqrt(var) : 0.0;
            double horizon = mean + 10.0 * sigma;
            if (!(horizon > 0.0) || Double.isInfinite(horizon) || Double.isNaN(horizon)) {
                horizon = 1.0;
            }
            for (int k = 0; k < Me_sample.MAX_DOUBLINGS; k++) {
                if (maxRowSurvival(horizon) < Me_sample.TAIL_MASS_TOL) {
                    break;
                }
                horizon *= 2.0;
            }

            this.gridSize = Me_sample.GRID_POINTS;
            this.h = horizon / (gridSize - 1);
            this.tGrid = new double[gridSize];
            this.eGrid = new double[gridSize][][];
            this.gGrid = new double[gridSize][];

            double[][] eh = Me_sample.toArray2D(D0.scale(h).expm_higham());
            double[][] identity = new double[n][n];
            for (int i = 0; i < n; i++) {
                identity[i][i] = 1.0;
            }
            eGrid[0] = identity;
            tGrid[0] = 0.0;
            gGrid[0] = rowSums(identity);
            for (int i = 1; i < gridSize; i++) {
                eGrid[i] = matTimesMat(eGrid[i - 1], eh);
                tGrid[i] = i * h;
                gGrid[i] = rowSums(eGrid[i]);
            }
            this.tEnd = tGrid[gridSize - 1];
            this.eta = Me_sample.dominantRate(D0, mean);
        }

        /** Largest row sum of expm(H0*t), an upper bound on any conditional survival. */
        private double maxRowSurvival(double t) {
            double[][] e = Me_sample.toArray2D(d0Matrix.scale(t).expm_higham());
            double[] rs = rowSums(e);
            double m = 0.0;
            for (int i = 0; i < n; i++) {
                m = Math.max(m, Math.abs(rs[i]));
            }
            return m;
        }

        /**
         * Draws one inter-event time and advances the conditional phase vector.
         *
         * @param random the uniform random source
         * @return the next inter-event time
         */
        public double next(Random random) {
            double u = random.nextDouble();
            if (exponential) {
                return -Math.log(1.0 - u) / expRate;
            }
            double target = 1.0 - u;

            double x;
            int lo;
            double sLast = Me_sample.dot(v, gGrid[gridSize - 1]);
            if (sLast >= target) {
                lo = gridSize - 1;
                if (sLast > 0.0 && target > 0.0 && eta < 0.0) {
                    x = tEnd + Math.log(sLast / target) / (-eta);
                    if (!(x > tEnd)) {
                        x = tEnd;
                    }
                } else {
                    x = tEnd;
                }
            } else {
                lo = 0;
                int hi = gridSize - 1;
                while (hi - lo > 1) {
                    int mid = (lo + hi) >>> 1;
                    if (Me_sample.dot(v, gGrid[mid]) >= target) {
                        lo = mid;
                    } else {
                        hi = mid;
                    }
                }
                double sLo = Me_sample.dot(v, gGrid[lo]);
                double sHi = Me_sample.dot(v, gGrid[lo + 1]);
                double denom = sLo - sHi;
                x = (denom > 0.0) ? tGrid[lo] + (sLo - target) / denom * h : tGrid[lo];
                double left = tGrid[lo];
                double right = left + h;
                double[] wLo = Me_sample.vecTimesMat(v, eGrid[lo]);
                for (int k = 0; k < Me_sample.NEWTON_STEPS; k++) {
                    double[] w = Me_sample.expmPropagate(wLo, d0, d0Matrix, normD0, x - left);
                    double surv = Me_sample.sum(w);
                    double f = -Me_sample.dot(w, rowSumD0);
                    if (!(f > 0.0)) {
                        break;
                    }
                    // Newton on g(x) = S(x) - target, with g'(x) = S'(x) = -f:
                    // x <- x - g/g' = x + (S(x) - target)/f. When the survival at
                    // x still exceeds the target the root lies to the right, so
                    // the step must be positive.
                    double err = surv - target;
                    if (Math.abs(err) < 1e-14) {
                        break;
                    }
                    double xn = x + err / f;
                    if (!(xn > left) || !(xn < right)) {
                        break;
                    }
                    boolean converged = Math.abs(xn - x) <= 1e-15 * Math.max(1.0, Math.abs(x));
                    x = xn;
                    if (converged) {
                        break;
                    }
                }
            }

            double[] wStart = Me_sample.vecTimesMat(v, eGrid[lo]);
            double[] w = Me_sample.expmPropagate(wStart, d0, d0Matrix, normD0, x - tGrid[lo]);
            double[] vn = Me_sample.vecTimesMat(w, d1);
            double norm = Me_sample.sum(vn);
            if (norm > 1e-300 && !Double.isNaN(norm) && !Double.isInfinite(norm)) {
                for (int j = 0; j < n; j++) {
                    vn[j] /= norm;
                }
                v = vn;
            } else {
                v = pie.clone();
            }
            return x;
        }

        /** Resets the conditional vector to the embedded stationary vector. */
        public void reset() {
            this.v = this.pie.clone();
        }
    }

    /**
     * Generates a correlated sample sequence from a Rational Arrival Process.
     *
     * <p>The returned inter-event times are <em>not</em> i.i.d.: they carry the
     * autocorrelation of the process, obtained by propagating the conditional
     * phase vector across events. A MAP is the special case in which H0 has
     * non-negative off-diagonal entries and H1 is non-negative.</p>
     *
     * @param RAP    the process as a MatrixCell {H0, H1}
     * @param n      the number of samples to generate
     * @param random the random number generator to use
     * @return an array of n correlated inter-event times
     */
    public static double[] rap_sample(MatrixCell RAP, long n, Random random) {
        return rap_sample(RAP.get(0), RAP.get(1), n, random);
    }

    /**
     * Generates a correlated sample sequence from a Rational Arrival Process
     * given its two matrices.
     *
     * @param H0     the hidden-transition matrix
     * @param H1     the visible-transition matrix
     * @param n      the number of samples to generate
     * @param random the random number generator to use
     * @return an array of n correlated inter-event times
     */
    public static double[] rap_sample(Matrix H0, Matrix H1, long n, Random random) {
        RapSampler sampler = new RapSampler(H0, H1);
        double[] samples = new double[(int) n];
        for (int i = 0; i < (int) n; i++) {
            samples[i] = sampler.next(random);
        }
        return samples;
    }

    /** Row sums of a square matrix. */
    private static double[] rowSums(double[][] m) {
        double[] out = new double[m.length];
        for (int i = 0; i < m.length; i++) {
            double s = 0.0;
            for (int j = 0; j < m[i].length; j++) {
                s += m[i][j];
            }
            out[i] = s;
        }
        return out;
    }

    /** Dense matrix product. */
    private static double[][] matTimesMat(double[][] x, double[][] y) {
        int rows = x.length;
        int inner = y.length;
        int cols = y[0].length;
        double[][] out = new double[rows][cols];
        for (int i = 0; i < rows; i++) {
            double[] xi = x[i];
            double[] oi = out[i];
            for (int k = 0; k < inner; k++) {
                double xik = xi[k];
                if (xik == 0.0) continue;
                double[] yk = y[k];
                for (int j = 0; j < cols; j++) {
                    oi[j] += xik * yk[j];
                }
            }
        }
        return out;
    }
}
